pkg/gardenlet/controller/federatedseed/extensions/controllerinstallation_control.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gardener/gardener/pkg/api/extensions"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type controllerInstallationControl struct {
+	k8sGardenClient kubernetes.Interface
+	seedClient      kubernetes.Interface
+	seedName        string
+	log             *logrus.Entry
+
+	lock              *sync.RWMutex
+	requiredKindTypes map[string]sets.String
+}
+
+func (c *controllerInstallationControl) createControllerInstallationReconcileFunc(ctx context.Context, kind string, newListObjFunc func() runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		// Step 1: List all existing extension objects of the given kind, collect the types and store it in the
+		// requiredKindTypes map.
+
+		listObj := newListObjFunc()
+		if err := c.seedClient.Client().List(ctx, listObj); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		requiredTypes := sets.NewString()
+		if err := meta.EachListItem(listObj, func(o runtime.Object) error {
+			obj, err := extensions.Accessor(o)
+			if err != nil {
+				return err
+			}
+
+			requiredTypes.Insert(obj.GetExtensionSpec().GetExtensionType())
+			return nil
+		}); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		c.lock.Lock()
+		c.requiredKindTypes[kind] = requiredTypes
+		c.lock.Unlock()
+
+		// Step 2: List all existing controller installation objects for the seed cluster this controller is responsible
```

Comment:
Hm, I didn't like the periodic idea too much as it exposes the risk of creating new race conditions. But you were bringing up a valid point which was that this reconciler was doing too much (not only handling the `kind` it was created for but all others as well).
I refactored it a little bit and changed it in the following way:

1. Compute all required types of existing extension objects for the reconciler `kind`.
1. List all `ControllerRegistration`s and store the names + extension types of those that are responsible for resources of the reconciler `kind`.
1. List all `ControllerInstallation`s, filter out those of different seeds and those that are referencing unrelated registrations (as of the previous step). Now only those `ControllerInstallation`s are handled that reference a registration that is responsible for the reconciler `kind`. If its supported extension types are part of the required list (computed in step 1) then set the `Required` condition to `True`, otherwise to `False`.

This way the reconcilers are only doing the job they were created for, and I also got rid of the `sync.RWMutex`.

Also, these loops are not happening too often (only if new extension objects are created/deleted or if the `.spec.type` changes (which never happens)).